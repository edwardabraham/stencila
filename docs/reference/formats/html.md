# HTML

<!-- prettier-ignore-start -->
<!-- CODEC-DOCS:START -->
<!-- CODEC-DOCS:STOP -->
<!-- prettier-ignore-end -->
