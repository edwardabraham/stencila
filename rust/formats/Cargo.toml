[package]
name = "formats"
description = "Metadata on file formats"
version = "0.0.0"
edition = "2021"

[dependencies]
eyre = "0.6.5"
once_cell = "1.8.0"
schemars = "0.8.6"
serde = "1.0.130"
serde_json = "1.0.68"
serde_with = "1.11.0"
