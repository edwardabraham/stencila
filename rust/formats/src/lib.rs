use eyre::Result;
use once_cell::sync::Lazy;
use schemars::{schema_for, JsonSchema};
use serde::Serialize;
use serde_json::json;
use serde_with::skip_serializing_none;
use std::{collections::HashMap, path::Path};

/// A list of known formats
#[rustfmt::skip]
pub static FORMATS: Lazy<Formats> = Lazy::new(|| {
    let formats = vec![
        // Data serialization formats. These may be used
        // to store documents so `preview: true`.
        Format::new("json", false, true, FormatNodeType::Unknown, &[]),
        Format::new("json5", false, true, FormatNodeType::Unknown, &[]),
        Format::new("toml", false, true, FormatNodeType::Unknown, &[]),
        Format::new("xml", false, true, FormatNodeType::Unknown, &[]),
        Format::new("yaml", false, true, FormatNodeType::Unknown, &[]),

        // Code formats
        Format::new("dockerfile", false, false, FormatNodeType::SoftwareSourceCode, &[]),
        Format::new("js", false, false, FormatNodeType::SoftwareSourceCode, &["javascript"]),
        Format::new("makefile", false, false, FormatNodeType::SoftwareSourceCode, &[]),
        Format::new("py", false, false, FormatNodeType::SoftwareSourceCode, &["python"]),
        Format::new("r", false, false, FormatNodeType::SoftwareSourceCode, &[]),
        Format::new("sh", false, false, FormatNodeType::SoftwareSourceCode, &["shell"]),
        Format::new("ts", false, false, FormatNodeType::SoftwareSourceCode, &["typescript"]),

        // Article formats
        Format::new("docx", true, true, FormatNodeType::Article, &[]),
        Format::new("html", false, true, FormatNodeType::Article, &[]),
        Format::new("ipynb", false, true, FormatNodeType::Article, &[]),
        Format::new("md", false, true, FormatNodeType::Article, &[]),
        Format::new("odt", true, true, FormatNodeType::Article, &[]),
        Format::new("rmd", false, true, FormatNodeType::Article, &[]),
        Format::new("latex", false, true, FormatNodeType::Article, &["tex"]),

        // Audio formats
        Format::new("flac", true, true, FormatNodeType::AudioObject, &[]),
        Format::new("mp3", true, true, FormatNodeType::AudioObject, &[]),
        Format::new("ogg", true, true, FormatNodeType::AudioObject, &[]),

        // Image formats
        Format::new("gif", true, true, FormatNodeType::ImageObject, &[]),
        Format::new("jpg", true, true, FormatNodeType::ImageObject, &["jpeg"]),
        Format::new("png", true, true, FormatNodeType::ImageObject, &[]),
        Format::new("rpng", true, true, FormatNodeType::ImageObject, &[]),

        // Video formats
        Format::new("3gp", true, true, FormatNodeType::VideoObject, &[]),
        Format::new("mp4", true, true, FormatNodeType::VideoObject, &[]),
        Format::new("ogv", true, true, FormatNodeType::VideoObject, &[]),
        Format::new("webm", true, true, FormatNodeType::VideoObject, &[]),

        // Other
        Format::new("txt", false, false, FormatNodeType::Unknown, &[]),

        // Specials
        Format::directory(),
    ];

    let formats = formats
        .into_iter()
        .map(|format| (format.name.clone(), format))
        .collect();

    Formats { formats }
});

/// The type of format as a schema `Node` type
#[derive(Clone, Debug, JsonSchema, Serialize)]
pub enum FormatNodeType {
    Article,
    AudioObject,
    ImageObject,
    VideoObject,
    SoftwareSourceCode,
    Unknown,
}

/// Information about a document format
///
/// Used to determine various application behaviors
/// e.g. not reading binary formats into memory unnecessarily
#[skip_serializing_none]
#[derive(Clone, Debug, JsonSchema, Serialize)]
#[schemars(deny_unknown_fields)]
pub struct Format {
    /// The lowercase name of the format e.g. `md`, `docx`, `dockerfile`
    pub name: String,

    /// Whether or not the format should be considered binary
    /// e.g. not to be displayed in a text / code editor
    pub binary: bool,

    /// Whether HTML previews are normally supported for documents of
    /// this format. See also `Document.previewable` which indicates whether
    /// a HTML preview is supported for a particular document.
    pub preview: bool,

    /// The kind of format
    pub node_type: FormatNodeType,

    /// Any additional extensions (other than it's name) that this format
    /// should match against.
    pub aliases: Vec<String>,

    /// Whether or not this is a known format (ie.e. not automatically created)
    pub known: bool,
}

impl Format {
    /// Create a new file format
    pub fn new(
        name: &str,
        binary: bool,
        preview: bool,
        kind: FormatNodeType,
        extensions: &[&str],
    ) -> Format {
        Format {
            known: true,
            name: name.into(),
            binary,
            preview,
            node_type: kind,
            aliases: extensions.iter().map(|s| s.to_string()).collect(),
        }
    }

    /// Create the special `directory` format used on `File` objects
    /// that are directories
    pub fn directory() -> Format {
        Format {
            known: true,
            name: "dir".into(),
            binary: true,
            preview: false,
            node_type: FormatNodeType::Unknown,
            aliases: Vec::new(),
        }
    }

    /// Create the special `unknown` file format where all we
    /// have is the name e.g. from a file extension.
    pub fn unknown(name: &str) -> Format {
        Format {
            known: false,
            name: name.into(),
            // Set binary to false so that any unregistered format
            // will be at least shown in editor...
            binary: false,
            // ..but not have a preview
            preview: false,
            node_type: FormatNodeType::Unknown,
            aliases: Vec::new(),
        }
    }
}

/// List of known document formats
#[derive(Serialize)]
#[serde(transparent)]
pub struct Formats {
    /// Document formats keyed by their name
    formats: HashMap<String, Format>,
}

impl Formats {
    /// Match a format name to a `Format`
    pub fn match_name(&self, name: &str) -> Format {
        match self.formats.get(&name.to_lowercase()) {
            Some(format) => format.clone(),
            None => {
                for format in self.formats.values() {
                    let name = name.to_string();
                    if format.aliases.contains(&name) {
                        return format.clone();
                    }
                }
                Format::unknown(name)
            }
        }
    }

    /// Match a file path to a `Format`
    pub fn match_path<P: AsRef<Path>>(&self, path: &P) -> Format {
        let path = path.as_ref();

        // Get name from file extension, or filename if no extension
        let name = match path.extension() {
            Some(ext) => ext,
            None => match path.file_name() {
                Some(name) => name,
                // Fallback to the provided "path"
                None => path.as_os_str(),
            },
        };

        // Match that name
        self.match_name(&name.to_string_lossy().to_string())
    }
}

/// Get JSON Schemas for this module
pub fn schemas() -> Result<serde_json::Value> {
    Ok(json!([schema_for!(Format)]))
}
