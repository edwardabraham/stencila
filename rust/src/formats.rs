use crate::utils::schemas;
use eyre::Result;
use once_cell::sync::Lazy;
use schemars::JsonSchema;
use serde::Serialize;
use serde_with::skip_serializing_none;
use std::{collections::HashMap, path::Path};

/// Information about a document format
///
/// Used to determine various application behaviors
/// e.g. not reading binary formats into memory unnecessarily
#[skip_serializing_none]
#[derive(Clone, Debug, JsonSchema, Serialize)]
#[schemars(deny_unknown_fields)]
pub struct Format {
    /// The lowercase name of the format e.g. `md`, `docx`, `dockerfile`
    pub name: String,

    /// Whether or not the format should be considered binary
    /// e.g. not to be displayed in a text / code editor
    pub binary: bool,

    /// The type of `CreativeWork` that this format is expected to be.
    /// This will be `None` for data serialization formats such as
    /// JSON or YAML which have no expected type (the actual type is
    /// embedded in the data).
    #[serde(rename = "type")]
    pub type_: Option<String>,
}

impl Format {
    /// Create a new file format
    pub fn new(name: &str, binary: bool, type_: &str) -> Format {
        Format {
            name: name.into(),
            binary,
            type_: if type_.is_empty() {
                None
            } else {
                Some(type_.to_string())
            },
        }
    }

    /// Create the special `unknown` file format
    pub fn unknown() -> Format {
        Format {
            name: "unknown".to_string(),
            binary: true,
            type_: None,
        }
    }
}

/// List of known document formats
#[derive(Serialize)]
#[serde(transparent)]
pub struct Formats {
    /// Document formats keyed by their name
    formats: HashMap<String, Format>,
}

impl Default for Formats {
    fn default() -> Formats {
        let formats = vec![
            // Data serialization formats
            Format::new("json", false, ""),
            Format::new("json5", false, ""),
            Format::new("toml", false, ""),
            Format::new("xml", false, ""),
            Format::new("yaml", false, ""),
            // Code formats
            Format::new("dockerfile", false, ""),
            Format::new("js", false, ""),
            Format::new("makefile", false, ""),
            Format::new("py", false, ""),
            Format::new("r", false, ""),
            Format::new("sh", false, ""),
            Format::new("ts", false, ""),
            // Article formats
            Format::new("docx", true, "Article"),
            Format::new("odt", true, "Article"),
            Format::new("ipynb", false, "Article"),
            Format::new("md", false, "Article"),
            Format::new("rmd", false, "Article"),
            Format::new("tex", false, "Article"),
            Format::new("txt", false, "Article"),
            // Audio formats
            Format::new("flac", true, "AudioObject"),
            Format::new("mp3", true, "AudioObject"),
            Format::new("ogg", true, "AudioObject"),
            // Image formats
            Format::new("gif", true, "ImageObject"),
            Format::new("jpeg", true, "ImageObject"),
            Format::new("jpg", true, "ImageObject"),
            Format::new("png", true, "ImageObject"),
            // Video formats
            Format::new("3gp", true, "VideoObject"),
            Format::new("mp4", true, "VideoObject"),
            Format::new("ogv", true, "VideoObject"),
            Format::new("webm", true, "VideoObject"),
            // Special `unknown` format which defines handling
            // of file types that are not registered above
            Format::unknown(),
        ];

        let formats = formats
            .into_iter()
            .map(|format| (format.name.clone(), format))
            .collect();

        Formats { formats }
    }
}

impl Formats {
    /// Match a format name to a `Format`
    pub fn match_name(&self, name: &str) -> Format {
        match self.formats.get(&name.to_lowercase()) {
            Some(format) => format.clone(),
            None => Format::unknown(),
        }
    }

    /// Match a file path to a `Format`
    pub fn match_path<P: AsRef<Path>>(&self, path: &P) -> Format {
        let path = path.as_ref();
        // Use file extension
        let name = match path.extension() {
            Some(ext) => ext,
            // Fallback to the filename if no extension
            None => match path.file_name() {
                Some(name) => name,
                // Fallback to the provided "path"
                None => path.as_os_str(),
            },
        };

        self.match_name(&name.to_string_lossy().to_string())
    }
}

pub static FORMATS: Lazy<Formats> = Lazy::new(Formats::default);

/// Get JSON Schemas for this modules
pub fn schemas() -> Result<serde_json::Value> {
    let schemas = serde_json::Value::Array(vec![schemas::generate::<Format>()?]);
    Ok(schemas)
}
