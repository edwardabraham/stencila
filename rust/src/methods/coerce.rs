use eyre::Result;
use stencila_schema::Node;

/// Coerce a JSON value to the Stencila schema
///
/// This function is intended to be used when deserializing
/// generic data formats e.g JSON, YAML prior to
/// avoid `serde` validation errors when subsequently deserializing
/// to a `Node`. Places this might be necessary:
///
/// - when decoding JSON, YAML, etc documents
/// - when deserializing the result from delegating a method
///   to a peer or plugin
/// - when decoding the YAML header of a Markdown document
pub fn coerce(value: serde_json::Value) -> Result<Node> {
    // TODO
    Ok(serde_json::from_value(value)?)
}
