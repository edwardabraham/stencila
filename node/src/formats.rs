use crate::prelude::*;
use neon::prelude::*;
use stencila::formats;

/// Get the module's schemas
pub fn schemas(cx: FunctionContext) -> JsResult<JsString> {
    let schemas = formats::schemas();
    to_json_or_throw(cx, schemas)
}

/// Get the hash map of file formats
pub fn formats(cx: FunctionContext) -> JsResult<JsString> {
    let formats = &*formats::FORMATS;
    to_json(cx, formats)
}
