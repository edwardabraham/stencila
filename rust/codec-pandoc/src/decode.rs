use crate::from_pandoc;
use codec_json::JsonCodec;
use codec_rpng::RpngCodec;
use codec_trait::{
    eyre::{bail, Result},
    stencila_schema::*,
    Codec,
};
use codec_txt::ToTxt;
use formats::{FormatNodeType, FORMATS};
use node_coerce::coerce;
use pandoc_types::definition as pandoc;
use std::{collections::HashMap, path::PathBuf};

/// Decode a document to a `Node`
///
/// Intended primarily for use by other internal codec crates e.g. `codec-docx`, `codec-latex`
pub async fn decode(
    input: &str,
    path: Option<PathBuf>,
    format: &str,
    args: &[&str],
) -> Result<Node> {
    let pandoc = from_pandoc(input, path, format, args).await?;
    decode_pandoc(pandoc)
}

/// Decode a fragment to a vector of `BlockContent`
///
/// Intended for decoding a fragment of a larger document (e.g. from some LaTeX in
/// a Markdown document). Ignores any meta data e.g. title
pub async fn decode_fragment(
    input: &str,
    format: &str,
    args: &[&str],
) -> Result<Vec<BlockContent>> {
    if input.is_empty() {
        return Ok(vec![]);
    }

    let pandoc = from_pandoc(input, None, format, args).await?;
    let context = DecodeContext {};
    Ok(translate_blocks(&pandoc.1, &context))
}

/// Decode a Pandoc document to a `Node`
pub fn decode_pandoc(pandoc: pandoc::Pandoc) -> Result<Node> {
    let context = DecodeContext {};
    let mut article = translate_meta(pandoc.0, &context)?;

    let content = translate_blocks(&pandoc.1, &context);
    article.content = if content.is_empty() {
        None
    } else {
        Some(content)
    };

    Ok(Node::Article(article))
}

/// Translate Pandoc meta data into an `Article` node
fn translate_meta(meta: pandoc::Meta, context: &DecodeContext) -> Result<Article> {
    let mut article = translate_meta_map(&meta.0, context);
    article.insert("type".to_string(), serde_json::json!("Article"));

    let node = coerce(serde_json::Value::Object(article), None)?;

    match node {
        Node::Article(article) => Ok(article),
        _ => bail!("Expected an article"),
    }
}

/// Translate a map of `MetaValue` to a map of `serde_json::Value`
fn translate_meta_map(
    map: &HashMap<String, pandoc::MetaValue>,
    context: &DecodeContext,
) -> serde_json::Map<String, serde_json::Value> {
    map.iter()
        .map(|(key, value)| (key.clone(), translate_meta_value(value, context)))
        .collect()
}

/// Translate a meta value to a `serde_json::Value`
fn translate_meta_value(value: &pandoc::MetaValue, context: &DecodeContext) -> serde_json::Value {
    match value {
        pandoc::MetaValue::MetaMap(map) => {
            serde_json::Value::Object(translate_meta_map(map, context))
        }
        pandoc::MetaValue::MetaList(vec) => serde_json::Value::Array(
            vec.iter()
                .map(|value| translate_meta_value(value, context))
                .collect(),
        ),
        pandoc::MetaValue::MetaBool(bool) => serde_json::Value::Bool(*bool),
        pandoc::MetaValue::MetaString(string) => serde_json::Value::String(string.clone()),
        pandoc::MetaValue::MetaInlines(inlines) => serde_json::Value::Array(
            translate_inlines(inlines, context)
                .iter()
                .map(|inline| serde_json::to_value(inline).expect("Can serialize to JSON value"))
                .collect(),
        ),
        pandoc::MetaValue::MetaBlocks(blocks) => serde_json::Value::Array(
            translate_blocks(blocks, context)
                .iter()
                .map(|block| serde_json::to_value(block).expect("Can serialize to JSON value"))
                .collect(),
        ),
    }
}

/// Decoding context
struct DecodeContext {}

/// Translate a vector of Pandoc `Block` elements to a vector of `BlockContent` nodes
fn translate_blocks(elements: &[pandoc::Block], context: &DecodeContext) -> Vec<BlockContent> {
    elements
        .iter()
        .flat_map(|child| translate_block(child, context))
        .collect()
}

/// Translate a Pandoc `Block` element into a zero or more `BlockContent` nodes
///
/// Will ignore elements that are dealt with by `translate_inline`
fn translate_block(element: &pandoc::Block, context: &DecodeContext) -> Vec<BlockContent> {
    match element {
        pandoc::Block::Header(depth, attrs, inlines) => {
            vec![BlockContent::Heading(Heading {
                id: get_id(attrs),
                depth: Some(*depth as u8),
                content: translate_inlines(inlines, context),
                ..Default::default()
            })]
        }

        pandoc::Block::Para(inlines) => {
            let content = translate_inlines(inlines, context);
            if content.len() == 1 {
                if let Some(translated) = try_code_chunk(&content[0]) {
                    return vec![translated];
                }
            }
            vec![BlockContent::Paragraph(Paragraph {
                content,
                ..Default::default()
            })]
        }

        pandoc::Block::BlockQuote(blocks) => {
            vec![BlockContent::QuoteBlock(QuoteBlock {
                content: translate_blocks(blocks, context),
                ..Default::default()
            })]
        }

        pandoc::Block::CodeBlock(attrs, text) => {
            let id = get_id(attrs);
            let programming_language = get_attr(attrs, "classes").map(Box::new);
            vec![BlockContent::CodeBlock(CodeBlock {
                id,
                programming_language,
                text: text.clone(),
                ..Default::default()
            })]
        }

        pandoc::Block::OrderedList(.., items) | pandoc::Block::BulletList(items) => {
            let order = Some(match element {
                pandoc::Block::OrderedList(..) => ListOrder::Ascending,
                _ => ListOrder::Unordered,
            });
            let items = items
                .iter()
                .map(|blocks| {
                    let blocks = translate_blocks(blocks, context);
                    let content = if blocks.is_empty() {
                        None
                    } else {
                        Some(ListItemContent::VecBlockContent(blocks))
                    };
                    ListItem {
                        content,
                        ..Default::default()
                    }
                })
                .collect();
            vec![BlockContent::List(List {
                items,
                order,
                ..Default::default()
            })]
        }
        pandoc::Block::DefinitionList(definitions) => {
            let items = definitions
                .iter()
                .map(|(term, definitions)| {
                    let term = vec![BlockContent::Paragraph(Paragraph {
                        content: translate_inlines(term, context),
                        ..Default::default()
                    })];
                    let definitions = definitions
                        .iter()
                        .flat_map(|blocks| translate_blocks(blocks, context))
                        .collect();
                    let blocks = [term, definitions].concat();
                    let content = if blocks.is_empty() {
                        None
                    } else {
                        Some(ListItemContent::VecBlockContent(blocks))
                    };
                    ListItem {
                        content,
                        ..Default::default()
                    }
                })
                .collect();
            vec![BlockContent::List(List {
                items,
                ..Default::default()
            })]
        }

        pandoc::Block::Table(attrs, caption, _column_specs, head, bodies, foot) => {
            let id = get_id(attrs);

            let caption = translate_blocks(&caption.1, context);
            let caption = match caption.is_empty() {
                true => None,
                false => Some(Box::new(TableCaption::VecBlockContent(caption))),
            };

            let head: Vec<TableRow> = head
                .1
                .iter()
                .map(|row| translate_row(row, context, Some(TableRowRowType::Header)))
                .collect();
            let body: Vec<TableRow> = bodies
                .iter()
                .flat_map(|body| {
                    let intermediate_head: Vec<TableRow> = body
                        .2
                        .iter()
                        .map(|row| translate_row(row, context, Some(TableRowRowType::Header)))
                        .collect();
                    let intermediate_body: Vec<TableRow> = body
                        .3
                        .iter()
                        .map(|row| translate_row(row, context, None))
                        .collect();
                    [intermediate_head, intermediate_body].concat()
                })
                .collect();
            let foot: Vec<TableRow> = foot
                .1
                .iter()
                .map(|row| translate_row(row, context, Some(TableRowRowType::Footer)))
                .collect();
            let rows = [head, body, foot].concat();

            vec![BlockContent::Table(TableSimple {
                rows,
                caption,
                id,
                ..Default::default()
            })]
        }

        pandoc::Block::HorizontalRule => vec![BlockContent::ThematicBreak(ThematicBreak {
            ..Default::default()
        })],

        pandoc::Block::RawBlock(_format, _content) => {
            // TODO: Attempt to decode raw content; skip if not decode-able
            vec![]
        }

        // A line block is "multiple non-breaking lines" so just flatten into a paragraph
        pandoc::Block::LineBlock(lines) => vec![BlockContent::Paragraph(Paragraph {
            content: lines
                .iter()
                .flat_map(|inlines| translate_inlines(inlines, context))
                .collect(),
            ..Default::default()
        })],

        // Element types not supported by Stencila but with child elements that are
        pandoc::Block::Div(_attrs, blocks) => translate_blocks(blocks, context),
        pandoc::Block::Plain(inlines) => vec![BlockContent::Paragraph(Paragraph {
            content: translate_inlines(inlines, context),
            ..Default::default()
        })],

        // Element types not supported by Stencila
        pandoc::Block::Null => vec![],
    }
}

/// Translate a Pandoc table row into a `TableRow`
fn translate_row(
    row: &pandoc::Row,
    context: &DecodeContext,
    row_type: Option<TableRowRowType>,
) -> TableRow {
    let cells = row
        .1
        .iter()
        .map(|cell| translate_cell(cell, context))
        .collect();
    TableRow {
        cells,
        row_type,
        ..Default::default()
    }
}

/// Translate a Pandoc table cell into a `TableCell`
///
/// Pandoc always returns blocks for table cells, for example, even for
/// a single number it returns a paragraph. This is somewhat heavy weight,
/// and inconsistent with other decoders in this repo, so here we unwrap
/// a single paragraph to inlines.
fn translate_cell(cell: &pandoc::Cell, context: &DecodeContext) -> TableCell {
    let pandoc::Cell(_attrs, _alignment, row_span, col_span, blocks) = cell;
    let blocks = translate_blocks(blocks, context);
    let content = match blocks.len() {
        0 => None,
        1 => match &blocks[0] {
            BlockContent::Paragraph(paragraph) => Some(TableCellContent::VecInlineContent(
                paragraph.content.clone(),
            )),
            _ => Some(TableCellContent::VecBlockContent(blocks)),
        },
        _ => Some(TableCellContent::VecBlockContent(blocks)),
    };
    let rowspan = match row_span {
        1 => None,
        _ => Some((*row_span) as u32),
    };
    let colspan = match col_span {
        1 => None,
        _ => Some((*col_span) as u32),
    };
    TableCell {
        colspan,
        content,
        rowspan,
        ..Default::default()
    }
}

/// Translate a vector of Pandoc `Inline` elements to a vector of `InlineContent` nodes
fn translate_inlines(elements: &[pandoc::Inline], context: &DecodeContext) -> Vec<InlineContent> {
    let mut inlines: Vec<InlineContent> = elements
        .iter()
        .flat_map(|child| translate_inline(child, context))
        .collect();

    let mut index = 1;
    while index < inlines.len() {
        let curr = inlines[index].clone();
        match (&mut inlines[index - 1], curr) {
            (InlineContent::String(prev), InlineContent::String(curr)) => {
                match curr.as_str() {
                    "\u{2029}" => prev.push(' '),
                    _ => prev.push_str(&curr),
                };
                inlines.remove(index);
            }
            _ => {
                index += 1;
            }
        }
    }

    inlines
}

/// Translate a Pandoc `Inline` element into a zero or more `InlineContent` nodes
fn translate_inline(element: &pandoc::Inline, context: &DecodeContext) -> Vec<InlineContent> {
    let inlines = match element {
        pandoc::Inline::Str(string) => vec![InlineContent::String(string.clone())],
        pandoc::Inline::Space => vec![InlineContent::String(" ".to_string())],
        pandoc::Inline::SoftBreak => vec![InlineContent::String("\u{2029}".to_string())],

        pandoc::Inline::Emph(inlines) => vec![InlineContent::Emphasis(Emphasis {
            content: translate_inlines(inlines, context),
            ..Default::default()
        })],
        pandoc::Inline::Underline(inlines) => {
            vec![InlineContent::NontextualAnnotation(NontextualAnnotation {
                content: translate_inlines(inlines, context),
                ..Default::default()
            })]
        }
        pandoc::Inline::Strong(inlines) => vec![InlineContent::Strong(Strong {
            content: translate_inlines(inlines, context),
            ..Default::default()
        })],
        pandoc::Inline::Strikeout(inlines) => vec![InlineContent::Delete(Delete {
            content: translate_inlines(inlines, context),
            ..Default::default()
        })],
        pandoc::Inline::Superscript(inlines) => vec![InlineContent::Superscript(Superscript {
            content: translate_inlines(inlines, context),
            ..Default::default()
        })],
        pandoc::Inline::Subscript(inlines) => vec![InlineContent::Subscript(Subscript {
            content: translate_inlines(inlines, context),
            ..Default::default()
        })],
        pandoc::Inline::Quoted(_quote_type, inlines) => vec![InlineContent::Quote(Quote {
            content: translate_inlines(inlines, context),
            ..Default::default()
        })],

        pandoc::Inline::Code(attrs, text) => {
            let id = get_id(attrs);
            vec![InlineContent::CodeFragment(CodeFragment {
                id,
                text: text.clone(),
                ..Default::default()
            })]
        }
        pandoc::Inline::Math(_math_type, text) => vec![InlineContent::MathFragment(MathFragment {
            text: text.clone(),
            ..Default::default()
        })],

        pandoc::Inline::Link(attrs, inlines, target) => {
            let pandoc::Target(url, title) = target;
            vec![InlineContent::Link(Link {
                id: get_id(attrs),
                target: url.clone(),
                title: get_string_prop(title),
                content: translate_inlines(inlines, context),
                ..Default::default()
            })]
        }
        pandoc::Inline::Image(attrs, inlines, target) => {
            let id = get_id(attrs);

            let caption = translate_inlines(inlines, context).to_txt();
            let caption = match caption.is_empty() {
                true => None,
                false => Some(Box::new(caption)),
            };

            let pandoc::Target(url, title) = target;
            let content_url = url.clone();
            let title = match title.is_empty() {
                true => None,
                false => Some(Box::new(CreativeWorkTitle::String(title.to_string()))),
            };

            match FORMATS.match_path(&content_url).node_type {
                FormatNodeType::AudioObject => {
                    vec![InlineContent::AudioObject(AudioObjectSimple {
                        content_url,
                        title,
                        id,
                        ..Default::default()
                    })]
                }
                FormatNodeType::VideoObject => {
                    vec![InlineContent::VideoObject(VideoObjectSimple {
                        content_url,
                        title,
                        id,
                        ..Default::default()
                    })]
                }
                _ => vec![InlineContent::ImageObject(ImageObjectSimple {
                    content_url,
                    title,
                    caption,
                    id,
                    ..Default::default()
                })],
            }
        }

        pandoc::Inline::Cite(citations, _inlines) => {
            let items: Vec<Cite> = citations
                .iter()
                .map(|citation| {
                    // TODO: Use inlines and prefix and suffix (consider using Vec<InlineContent> for both
                    // and parsing out pagination)
                    let citation_mode = Some(match citation.citation_mode {
                        pandoc::CitationMode::NormalCitation => CiteCitationMode::Parenthetical,
                        pandoc::CitationMode::AuthorInText => CiteCitationMode::Narrative,
                        pandoc::CitationMode::SuppressAuthor => CiteCitationMode::NarrativeYear,
                    });
                    Cite {
                        citation_mode,
                        target: citation.citation_id.clone(),
                        ..Default::default()
                    }
                })
                .collect();
            match items.len() {
                0 => vec![],
                1 => vec![InlineContent::Cite(items[0].clone())],
                _ => vec![InlineContent::CiteGroup(CiteGroup {
                    items,
                    ..Default::default()
                })],
            }
        }

        pandoc::Inline::Note(_blocks) => {
            // TODO: Decode blocks to the context and link to their Stencila `Note`
            vec![]
        }

        pandoc::Inline::RawInline(_format, _content) => {
            // TODO: Attempt to decode raw content; skip if not decode-able
            vec![]
        }

        // Element types not supported by Stencila but with child elements that are
        pandoc::Inline::SmallCaps(inlines) | pandoc::Inline::Span(.., inlines) => {
            translate_inlines(inlines, context)
        }

        // Element types not supported by Stencila
        pandoc::Inline::LineBreak => vec![],
    };

    // Try to transform inline nodes as needed
    inlines
        .into_iter()
        .map(|inline| {
            if let Some(code_expression) = try_code_expression(&inline) {
                code_expression
            } else {
                inline
            }
        })
        .collect()
}

/// Get an attribute from a Pandoc `Attr` tuple struct
fn get_attr(attrs: &pandoc::Attr, name: &str) -> Option<String> {
    match name {
        "id" => match attrs.0.is_empty() {
            true => None,
            false => Some(attrs.0.clone()),
        },
        "classes" => match attrs.1.is_empty() {
            true => None,
            false => Some(attrs.1.join(" ")),
        },
        _ => attrs.2.iter().find_map(|(key, value)| {
            if key == name {
                Some(value.clone())
            } else {
                None
            }
        }),
    }
}

/// Get an optional string property from a string
/// If the string is empty, then will be `None`
fn get_string_prop(value: &str) -> Option<Box<String>> {
    match value.is_empty() {
        true => None,
        false => Some(Box::new(value.to_string())),
    }
}

// Get the `id` property of a `Entity` from a  Pandoc `Attr` tuple struct
fn get_id(attrs: &pandoc::Attr) -> Option<Box<String>> {
    get_attr(attrs, "id").and_then(|value| get_string_prop(&value))
}

/// Try to extract a `CodeExpression` from an RPNG representation
fn try_code_expression(inline: &InlineContent) -> Option<InlineContent> {
    match inline {
        InlineContent::Link(link) => {
            if link.content.len() == 1 {
                // If this is a link around a code expression and the it has a
                // matching title then return that code expression
                let title = match link.title.as_deref() {
                    Some(title) => title.clone(),
                    None => "".to_string(),
                };
                if title == "CodeExpression"
                    || link.target.starts_with("https://hub.stenci.la/api/nodes")
                {
                    if let InlineContent::CodeExpression(expr) = &link.content[0] {
                        return Some(InlineContent::CodeExpression(expr.clone()));
                    }
                }
                // Try to get a code expression from the inner content
                if let Some(expr) = try_code_expression(&link.content[0]) {
                    return Some(expr);
                }
            }
            // Fallback to fetching code expression from the link's URL
            // TODO
        }
        InlineContent::ImageObject(image) => {
            // Try to get the code expression from the caption
            if let Some(caption) = image.caption.as_deref() {
                if let Ok(Node::CodeExpression(expr)) = JsonCodec::from_str(caption, None) {
                    return Some(InlineContent::CodeExpression(expr));
                }
            }
            // Fallback to getting from the image
            if let Ok(Node::CodeExpression(expr)) = RpngCodec::from_str(&image.content_url, None) {
                return Some(InlineContent::CodeExpression(expr));
            }
        }
        _ => (),
    };
    None
}

/// Try to extract a `CodeChunk` from an RPNG representation
fn try_code_chunk(inline: &InlineContent) -> Option<BlockContent> {
    match inline {
        InlineContent::Link(link) => {
            // Try to get a code chunk from the inner content
            if link.content.len() == 1 {
                if let Some(chunk) = try_code_chunk(&link.content[0]) {
                    return Some(chunk);
                }
            }
            // Fallback to fetching code chunk from the link's URL
            // TODO
        }
        InlineContent::ImageObject(image) => {
            // Try to get the code chunk from the caption
            if let Some(caption) = image.caption.as_deref() {
                if let Ok(Node::CodeChunk(chunk)) = JsonCodec::from_str(caption, None) {
                    return Some(BlockContent::CodeChunk(chunk));
                }
            }
            // Fallback to getting from the image
            if let Ok(Node::CodeChunk(chunk)) = RpngCodec::from_str(&image.content_url, None) {
                return Some(BlockContent::CodeChunk(chunk));
            }
        }
        _ => (),
    };
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_snaps::{insta::assert_json_snapshot, snapshot_fixtures_content};

    #[test]
    fn pandoc_fragments() {
        let runtime = tokio::runtime::Runtime::new().unwrap();
        snapshot_fixtures_content("fragments/pandoc/*.json", |content| {
            let json =
                runtime.block_on(async { decode_fragment(content, "pandoc", &[]).await.unwrap() });
            assert_json_snapshot!(json);
        });
    }
}
