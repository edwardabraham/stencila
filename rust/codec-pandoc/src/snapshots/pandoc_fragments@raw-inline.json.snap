---
source: rust/codec-pandoc/src/decode.rs
expression: json
input_file: fixtures/fragments/pandoc/raw-inline.json

---
[
  {
    "type": "Paragraph",
    "content": [
      "This paragraph has some  inline."
    ]
  }
]
